### PR TITLE
Add Absolute Position Encoder

### LINKED ISSUE #4: Add PV/Unified Position encoding
## Detailed Description
As mentioned in https://github.com/openclimatefix/satflow/issues/101 it would be helpful to have a way of having consistent position encodings for the PV systems/satellite imagery so that the model can associate the PV systems output in time and space with the satellite imagery. 

This unified position encoding could also be useful for other modalities, so that the model only needs one set of position encodings but can use it for all the input modalities.

## Context
It would help with the joint model and unifying the inputs/outputs. This position encoding can also be used with the https://github.com/openclimatefix/perceiver-pytorch/issues/20 to ensure the queries use the same positional encodings when getting the output

## Possible Implementation
Could be just a way of encoding all of this with Fourier Features. Or could be some other way of encoding that works across the different modalities

### PR DESCRIPTION
**Pull request by jacobbieker**
# Pull Request

## Description

Moved from https://github.com/openclimatefix/nowcasting_utils/pull/33 

Adds positional encoders for:

1. Absolute Position Encodings

It is generated using Fourier Features for now, although other options could include coordinates, such as CoordConv or others.

Fixes #4 

## How Has This Been Tested?

Unit Tests, which pass for the encoding. Some fail for the dataloader/datamodule, but that's unrelated to this

- [ ] No
- [x] Yes

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/nowcasting/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: nowcasting_dataloader/utils/position_encoding.py [hunk 1]
```diff
@@ -153,7 +139,7 @@ def encode_absolute_position(
 
 
 def normalize_geospatial_coordinates(
-    geospatial_coordinates, geospatial_bounds: Dict[str, float], **kwargs
+    geospatial_coordinates: List[np.ndarray], geospatial_bounds: Dict[str, float], **kwargs
```
Comment (flowirtz) [inline nowcasting_dataloader/utils/position_encoding.py:142]: yay, more types!

File: nowcasting_dataloader/utils/position_encoding.py [hunk 2]
```diff
@@ -86,8 +75,8 @@ def encode_position(
         shape: The shape of the input to be encoded, should be the largest or finest-grained input
             For example, if the inputs are shapes (12, 6, 128, 128) and (1, 6), (12, 6, 128, 128) should be passed in as
             shape, as it has the most elements and the input (1, 6) can just subselect the position encoding
-        geospatial_coordinates: The latitude/longitude of the inputs for shape, in OSGB coordinates, unused if using relative positioning only
-        datetimes: time of day and date for each of the timesteps in the shape, unused if using relative positioning only
+        geospatial_coordinates: The latitude/longitude of the inputs for shape, in OSGB coordinates
+        datetimes: time of day and date for each of the timesteps in the shape
         method: Method of the encoding, either 'fourier' for Fourier Features
         positioning: The type of positioning used, either 'relative' for relative positioning, or 'absolute', or 'both'
```
Comment (flowirtz) [inline nowcasting_dataloader/utils/position_encoding.py:81]: remove from docstring as well
```suggestion
```

File: tests/test_position_encoding.py [hunk 3]
```diff
@@ -220,26 +216,14 @@ def test_encode_multiple_modalities():
...
-def test_not_implemented_option(positioning):
+def test_fake_method_option():
     datetimes, geospatial_bounds, geospatial_coordinates = get_data()
-    with pytest.raises(NotImplementedError):
-        encode_modalities(
-            modalities_to_encode={"NWP": torch.randn(16, 1, 1, 1)},
+    with pytest.raises(AssertionError):
+        encode_position(
+            shape=[1, 1, 1, 1],
+            positioning="relative",
```
Comment (flowirtz) [inline tests/test_position_encoding.py:224]: I think `positioning` should be removed here as well, as `method="test_method"` is supposed to trigger the `AssertionError` - or am I missing something?

```suggestion
```

Comment (jacobbieker) [reply tests/test_position_encoding.py:224]: Yes, you are right!

**Comment by jacobbieker** [comment]
> > Few smaller comments, but after fixing that LGTM! +1
> > > Some [tests] fail for the dataloader/datamodule
> > 
> > 
> > Are we tracking those anywhere already?
> 
> No, not right now, it seems like some changes in nowcasting-dataset meant the dataloading code I copied from SatFlow fails. Since nowcasting-dataset is still changing quite a bit with [openclimatefix/nowcasting_dataset#213](https://github.com/openclimatefix/nowcasting_dataset/issues/213) I was going with sorta waiting till that's more finished to fix the tests. I'll make pytest skip them actually for now

I'll be making these tests pass as part of #7

File: nowcasting_dataloader/utils/position_encoding.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+    """
+    datetime_features = create_datetime_features(datetimes)
+
+    # Fourier Features of absolute position
+    encoded_latlon = normalize_geospatial_coordinates(
+        geospatial_coordinates, geospatial_bounds, **kwargs
+    )
+
+    # Combine time and space features
+    to_concat = [einops.repeat(encoded_latlon, "b h w c -> b c t h w", t=shape[1])]
```
Comment (JackKelly) [inline nowcasting_dataloader/utils/position_encoding.py:128]: I wonder if, before we have named tensors, we could use consts to identify each dimension?  So, for example, instead of doing `t=shape[1]`, instead we could do:

 ```python
TIME_DIM = 1
# ...
t=shape[TIME_DIM]
```

Maybe?

Or an Enum?

I don't know about you but I easily forget which dimension is which!

Although I don't remember seeing things like `shape[TIME_DIM]` in other peoples' PyTorch code, so maybe it's a bad idea?!

File: nowcasting_dataloader/utils/position_encoding.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    return absolute_position_encoding
+
+
+def normalize_geospatial_coordinates(
+    geospatial_coordinates: List[np.ndarray], geospatial_bounds: Dict[str, float], **kwargs
+) -> torch.Tensor:
+    """
+    Normalize the geospatial coordinates by the max extant to keep everything between -1 and 1, in sin and cos
+
+    This should work on a batch level, as the geospatial bounds should be the same for every example in the batch
```
Comment (JackKelly) [inline nowcasting_dataloader/utils/position_encoding.py:146]: Ooooh... interesting....

I wonder if, actually, the position encoding (for both time and space) should be normalised so, for a given "real" location on Earth, the ML model always sees exactly the same position encoding??  My thinking is that the model might find it useful to learn that, say, clouds in Cornwall behave slightly differently to clouds in Scotland?

Having said all that, I think the code already works in this way?!  So maybe all that needs to be done here is maybe slightly clarify the docstring?

File: nowcasting_dataloader/utils/position_encoding.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+    for idx in range(len(geospatial_coordinates[0])):
+        x = geospatial_coordinates[0][idx]
+        y = geospatial_coordinates[1][idx]
+        grid = torch.meshgrid(x, y)
+        pos = torch.stack(grid, dim=-1)
+        encoded_position = fourier_encode(pos, **kwargs)
+        encoded_position = einops.rearrange(encoded_position, "... n d -> ... (n d)")
+        to_concat.append(encoded_position)
+
+    # And now convert to Fourier features, based off the absolute positions of the coordinates
```
Comment (JackKelly) [inline nowcasting_dataloader/utils/position_encoding.py:180]: Is this comment still relevant?

File: nowcasting_dataloader/utils/position_encoding.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+        radians = index * 2 * np.pi
+        index_sin = torch.sin(radians)
+        index_cos = torch.cos(radians)
+        outputs.append(index_sin)
+        outputs.append(index_cos)
+
+    return outputs
+
+
+def encode_fouier_position(
```
Comment (JackKelly) [inline nowcasting_dataloader/utils/position_encoding.py:220]: Tiny typo :slightly_smiling_face: 

Also, is this function called anywhere?

```suggestion
def encode_fourier_position(
```

File: nowcasting_dataloader/utils/position_encoding.py [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+    instead of appending the position encoding in case the position encoding needs to be used for the query to the
+    Perceiver IO model
+
+    This code assumes that there is at least 2 timesteps of at least one modality to be encoded
+
+    Args:
+        positioning: The type of positioning used, either 'relative' for relative positioning, or 'absolute', or 'both'
+        modalities_to_encode: Dict of input modalities, i.e. NWP, Satellite, PV, GSP, etc as torch.Tensors in [B, C, T, H, W] ordering
+        datetimes: Dict of datetimes for each modality, giving the actual date for each timestep in the modality
+        geospatial_coordinates: Dict of lat/lon coordinates for each modality with pixels, used to determine smallest spatial step needed, in OSGB coordinates
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply nowcasting_dataloader/utils/position_encoding.py:38]: Yeah, sounds good!

File: nowcasting_dataloader/utils/position_encoding.py [hunk 9]
```diff
@@ -0,0 +1,291 @@
+"""
+This file contains various ways of performing positional encoding.
+
+These encodings can be:
+- Relative positioning (i.e. this pixel is this far from the top left, and this many timesteps in the future)
```
[the remark this answers is not in the document]
Comment (jacobbieker) [reply nowcasting_dataloader/utils/position_encoding.py]: I think delete it, as we talked about, I think absolute encoding is the strictly better option in terms of what information it encodes, and that it is still relative, so I'll delete the line

File: nowcasting_dataloader/utils/position_encoding.py (hunk 7 above)
Comment (jacobbieker) [reply nowcasting_dataloader/utils/position_encoding.py:220]: No! It was for the relative encoding, but not now, removed it

File: nowcasting_dataloader/utils/position_encoding.py (hunk 5 above)
Comment (JackKelly) [reply nowcasting_dataloader/utils/position_encoding.py:115]: That all sounds good to me!  I think the code is great as it is.  Maybe all that's required is just tweaking the docstring (I read the docstring as meaning that the normalisation would be done on a batch-by-batch basis) if that's OK?
